### PR TITLE
GCP Observability Interop Testing Client/Server for Java

### PR DESCRIPTION
**Pull request by stanley-cheung**
This PR updates the existing interop test client/server so that they can be used by the main [GCP Observability Integration Interop testing driver](https://github.com/GoogleCloudPlatform/grpc-gcp-tools/pull/18).

We are re-using the existing interop test client and server as much as possible. Changes:
- Added an `--enable_observability=boolean` flag to both the client and the server. If enabled, we call `GcpObservability.grpcInit()` and `GcpObservability.close()`
- Added a `--num_times=INT` flag to the client to allow us to run one test case multiple times. 

These changes will be done across all supported languages

In addition to interop test, we added a `Dockerfile` and a `run.sh` script to allow the observability test driver to launch the interop client / server and accept parameters.

---

Original PR comment below (almost all info out-of-dated)

Key files to review:
- ~~[`ObservabilityTestClient.java`](9858/files#diff-7a0ea6be50d6334f67fea21a6efb66c60c7037b7a59db56f1c291ad6fb7cf910): This adds a new interop client while trying to re-use the existing interop testing logic as much as possible. I created a separate binary for this because it will make controlling the command line parameters and the action to take (i.e. details of making a particular RPC) a bit easier to manage.~~
- ~~[`ObservabilityTestServer.java`](9858/files#diff-7e1d7502f052349aa0f526e79fdb6dda217430743c82a7c9c74235d8973cb25d): This add a new interop server while completely reusing the existing interop server / service logic.~~
- [`build.gradle`](9858/files#diff-4402770413df55b9edf870b307904c4a7cce2e4012da42caea7ae790aa08d4fb): please let me know if I am doing the right thing here. This allows me to build the above binaries with this command: `./gradlew installDist -x test -PskipCodegen=true -PskipAndroid=true`.
- ~~[`ObservabilityTestUtil.java`](9858/files#diff-dc1b2481c289582067c3d6bf2e42fc11cd9949dddb368a412ac24ab5eb940edf): just some constants that may go away after I refactor the whole system a bit more across language.~~

~~There is a new [`Dockerfile`](9858/files#diff-97649f62a078e14bec3c244c94cdc9ab0ad6835087116d926a4d53b75826e404) and a [`build_docker.sh`](9858/files#diff-fc24ded0e3e0bc14887baaa55cc4ca21b605098043ed8947794d1ccfdc97d0eb) script under the `buildscripts` directory because some of the tests require testing in GKE and we need to deploy a docker image there. For now I just try to copy the binaries directly into the docker image and execute them. This might change in the future.~~

### CODE REVIEW

**Review by larry-safran — COMMENTED** [review]
I don't see anywhere that checks that observability information is actually being passed anywhere.  I may be mistaken, but it seems to be a simplification of the pingpong test. 

exportInterval doesn't appear to be used to actually do an export.  

The while loop at the end of the client seems aimed at interactive running rather than an automated interop test.

**Comment by stanley-cheung** [comment]
Thanks for the review

> I don't see anywhere that checks that observability information is actually being passed anywhere. I may be mistaken, but it seems to be a simplification of the pingpong test.

So the only public API is to "enable" Observability [here](9858/files#diff-7a0ea6be50d6334f67fea21a6efb66c60c7037b7a59db56f1c291ad6fb7cf910R148). The observability config is being supplied from the environment and being managed by the main test driver outside of this.

> 
> exportInterval doesn't appear to be used to actually do an export.

Yea the naming here isn't ideal. Ideally, and eventually, this variable should go away in the future. This variable currently exists just because we want to wait for the underlying opencensus module to have enough time to flush / export their metrics/traces to cloud. It's basically a hack at this point.

> 
> The while loop at the end of the client seems aimed at interactive running rather than an automated interop test.

Yes this is a relic from a time where I want to make sure things are sleeping properly. It's not needed for the automated interop test.

**Comment by larry-safran** [comment]
> Yea the naming here isn't ideal. Ideally, and eventually, this variable should go away in the future. This variable currently exists just because we want to wait for the underlying opencensus module to have enough time to flush / export their metrics/traces to cloud. It's basically a hack at this point.

It would be good to put in a comment capturing the explanation.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java:262]: Don't have this at all. This should be controlled by this implementation, not the test framework. This code has much better knowledge of the upload frequency than the test framework. Just hard-code the appropriate sleep into the code when observability is enabled.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java:295]: There's no benefit to special-casing non-`,`-containing strings. The comma code can run in either case.

**Review by ejona86 — COMMENTED** [review]
This looks good. Only real concern is redefining `--test_case`. That is defined cross-language at https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md#client

We will wait for merging until the test framework is merged.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (stanley-cheung) [reply interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java]: Implemented. Also remove that from the framework and other lang's implementation. Right now it's hardcoded at 65 seconds everywhere.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (stanley-cheung) [reply interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java]: Fixed.

**Comment by stanley-cheung** [comment]
@ejona86 As discussed, I have removed the handling of comma delimited `--test_case`. Instead, I added the `--num_times` parameter to the interop client. This has been done to all the other languages too.

From the main test driver, `large_unary,ping_pong` will now be 2 separate parallel processes. `large_unary` 20 times will instead be `--test_case=large_unary --num_times=20`.

Btw, I seem to have broken some tests:

```
> Task :grpc-servlet-jakarta:tomcat10Test

io.grpc.servlet.jakarta.TomcatInteropTest > veryLargeRequest FAILED
    java.lang.ClassCastException at TomcatInteropTest.java:100
```

Any ideas? I don't seem to have touched the `AbstractInteropTest` class...

**Comment by ejona86** [comment]
You changed the dependencies, which could cause ManagedChannelBuilder to return a different type. But the code doesn't need to work like that; TomcatInteropTest should just use `NettyChannelBuilder.forAddress()` instead of ManagedChannelBuilder.

**Comment by stanley-cheung** [comment]
All PR tests passed!

File: buildscripts/observability-test/Dockerfile
Comment (ejona86) [inline buildscripts/observability-test/Dockerfile:1]: 2023. In the shell scripts as well.

File: buildscripts/observability-test/run.sh
Comment (ejona86) [inline buildscripts/observability-test/run.sh:19]: Talked offline, and I do really want this. The test framework would pass in `--server_host=blah` and such as arguments, and this script would handle any number of arguments (beyond the client/server first one). But that change will happen after this is initially merged.

**Review by ejona86 — APPROVED** [review]

File: buildscripts/observability-test/Dockerfile
Comment (stanley-cheung) [reply buildscripts/observability-test/Dockerfile]: Done. Damn I started working on this since 2022 :)

**Review by larry-safran — APPROVED** [review]

**Pull request merged**
